INT-3637: JMX Improvements Msg Sources/Handlers

garyrussell opened the pull request:
JIRA: https://jira.spring.io/browse/INT-3637
- Message Sources
- Message Handlers

garyrussell committed:
INT-3637: JMX Improvements Msg Sources/Handlers

JIRA: https://jira.spring.io/browse/INT-3637

- Message Sources
- Message Handlers

garyrussell commented:
Review only - I need to do my own final review (it's late). There are also a couple of minor TODOs.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/config/annotation/AbstractMethodAnnotationPostProcessor.java:
```diff
@@ -150,7 +150,7 @@ public Object postProcess(Object bean, String beanName, Method method, List<Anno
 				}
 				else {
-					ProxyFactory proxyFactory = new ProxyFactory(bean);
+					ProxyFactory proxyFactory = new ProxyFactory(handler);
```
Ah! Must be backported

artembilan replied:
BTW, how have you catched it? Do you any specific test? 
I mean save the time for the backporting :smile:

garyrussell replied:
I hit it while testing `IdempotentReceiverIntegrationTests` (after I removed the JMX proxy) - since the test is in JMX, it always took the other branch.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/support/management/MetricsEnablement.java:
```diff
@@ -0,0 +1,37 @@
+	@ManagedAttribute
+	boolean isStatsEnabled();
+
+}
```
New line in the end of class

artembilan commented on spring-integration-jmx/src/test/java/org/springframework/integration/jmx/config/DynamicRouterTests.java:
```diff
@@ -109,4 +112,12 @@ public void testRouteChangeMapNamedArgs() throws Exception {
+			this.nullChannel.send(null);
+		}
+		System.out.println("done");
+	}
```
Redundant test.
Or with `@Ignore` and comment respectively

garyrussell committed:
Polishing; Final Review and PR Comments

- avoid second call to System.currentTimeMillis()
- fix elapsed time in handler metrics
- expose TrackableComponent when available

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/channel/NullChannel.java:
```diff
@@ -67,12 +69,31 @@ public String getComponentType() {
+			this.countsEnabled = true;
+		}
 		this.statsEnabled = statsEnabled;
+		if (this.channelMetrics != null) {
```
Redundant `null check`

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/channel/management/ChannelSendMetrics.java:
```diff
@@ -79,33 +83,38 @@ public String getName() {
-			if (result) {
+	public void afterSend(long start, boolean result) {
+		if (start > 0) {
+			long now = System.currentTimeMillis();
```
Looks like this `now` should be withing `if (result && this.fullStatsEnabled)` as well.

artembilan commented on spring-integration-jmx/src/main/java/org/springframework/integration/jmx/config/MBeanExporterHelper.java:
```diff
@@ -92,6 +92,20 @@ public void afterPropertiesSet() throws Exception {
+									|| className.endsWith("MessageHandler")
+									|| className.endsWith("MessageHandlerChain")
+									|| className.endsWith("BridgeHandler"))) { //TODO: more handlers?
+						// can't think of a better way - any ideas?
```
Explain, please, what is the goal we are pursuing here?

garyrussell replied:
I am trying to maintain backwards compatibility. Now that all components are `@ManagedResource` they will be picked up by a normal context mbean exporter. This code suppresses their inclusion (requiring an IMBE as before).

artembilan commented on spring-integration-jmx/src/main/java/org/springframework/integration/monitor/IntegrationMBeanExporter.java:
```diff
@@ -236,8 +237,7 @@ public Object postProcessAfterInitialization(Object bean, String beanName) throw
 		if (bean instanceof Advised) {
```
Looks like this `if` is redundant at all already.

garyrussell replied:
Yes; I was removing them one at a time; forgot the final removal.

artembilan commented on spring-integration-jmx/src/main/java/org/springframework/integration/monitor/IntegrationMBeanExporter.java:
```diff
@@ -250,24 +250,22 @@ public Object postProcessAfterInitialization(Object bean, String beanName) throw
 			}
-			SimpleMessageHandlerMetrics monitor = new SimpleMessageHandlerMetrics((MessageHandler) bean);
-			Object advised = applyHandlerInterceptor(bean, monitor, beanClassLoader);
+			MessageHandlerMetrics monitor = (MessageHandlerMetrics) extractTarget(bean);
```
Since all `Metrics` aren't already `Advices` there is no reason to invoke that `extractTarget()`.

garyrussell replied:
Good point - I was trying to be safe in case something else had advised it, but since we only reference interfaces, it's not necessary.

artembilan commented on spring-integration-jmx/src/main/java/org/springframework/integration/monitor/IntegrationMBeanExporter.java:
```diff
@@ -876,14 +834,14 @@ private String getChannelBeanKey(String channel) {
 	}
 
-	private String getSourceBeanKey(MessageSourceMetrics handler) {
+	private String getSourceBeanKey(MessageSourceMetrics monitor) {
```
s/monitor/source ?

artembilan commented on spring-integration-jmx/src/test/java/org/springframework/integration/jmx/UpdateMappingsTests.java:
```diff
@@ -95,7 +95,7 @@ public void testJmx() throws Exception {
 		messagingTemplate.setReceiveTimeout(1000);
 		Set<ObjectName> names = this.server.queryNames(ObjectName
-				.getInstance("update.mapping.domain:type=HeaderValueRouter,name=router"),
+				.getInstance("update.mapping.domain:type=MessageHandler,name=router,bean=endpoint"),
```
So, don't we have already the change to use the real class name for the `type` ?
Seems for it is enough useful from JMX tree.

garyrussell replied:
Previously, two MBeans were registered for the router. Now there's only one (with all operations).

artembilan commented:
Travis voice:
> org.springframework.integration.gateway.AsyncGatewayTests > promiseWithConsumer FAILED
java.lang.AssertionError at AsyncGatewayTests.java:345

Seems for me we should remove all those timing checks.
New JIRA? Or can you cover that here ?

garyrussell commented:
100ms is way too short but I don't think we should remove the test completely.

garyrussell committed:
INT-3637: Add @IntergrationManagedResource

Prevents beans (channels, etc), which were previously picked
up via a proxy, from being exported by a standard context
MBeanExporter.

I looked at completely eliminating the MBeanExporterHelper,
which suppresses o.s.integration beans from being exported by
a standard MBeanExporter when there is an IMBE present, but I feel
this is too much of a breaking change. There are a number of
standard beans (such as WireTap) that are @ManagedResources and
these would disappear for users that don't have an IMBE.

That said, such beans previously disappear completely when there
*is* an IMBE so now they are now annotated with both so that
they are exported by at most one of the exporters.

garyrussell commented:
Rebased; pushed:

__INT-3637: Add @IntergrationManagedResource__

Prevents beans (channels, etc), which were previously picked
up via a proxy, from being exported by a standard context
`MBeanExporter`.

I looked at completely eliminating the `MBeanExporterHelper`,
which suppresses `o.s.integration beans` from being exported by
a standard `MBeanExporter` when there is an `IMBE` present, but I feel
this is too much of a breaking change. There are a number of
standard beans (such as `WireTap`) that are `@ManagedResources` and
these would disappear for users that don't have an `IMBE`.

That said, such beans previously disappear completely when there
*is* an `IMBE` so now they are now annotated with both so that
they are exported by at least/most one of the exporters.

garyrussell committed:
Polishing; Use MBE.addExludedBean instead of DFA

garyrussell commented:
More polish.

garyrussell committed:
INT-3639: JMX (AMQP/JMS) Channel Stats

JIRA: https://jira.spring.io/browse/INT-3639

Module channels inherit `ChannelSendMetrics`; add
`PollableChannelManagement` for polled channels.

garyrussell commented:
Added AMQP/JMS Channel support for new JMX stats

garyrussell committed:
INT-3638: JMX Initial Stats/Counts Settings

JIRA: https://jira.spring.io/browse/INT-3638

Add the ability to specify the initial settings for
'enableStats' and 'enableCounts' for MBeans that
support those statistics.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/channel/AbstractPollableChannel.java:
```diff
@@ -104,7 +105,7 @@ public long getReceiveErrorCountLong() {
-			if (isStatsEnabled()) {
+			if (isCountsEnabled()) {
 				getMetrics().afterReceive();
 			}
```
Missed `counted = true;` as you did in the `PollableAmqpChannel`

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/channel/interceptor/WireTap.java:
```diff
@@ -39,6 +40,7 @@
  * @author Gary Russell
  */
 @ManagedResource
+@IntegrationManagedResource
```
Both to cover the case when there is no `IME`, but the core exists ?

garyrussell replied:
Yes; I explained it in the commit comment (and added it as a comment here when I pushed).

>... There are a number of
standard beans (such as WireTap) that are @ManagedResources and
these would disappear for users that don't have an IMBE.

>That said, such beans previously disappear completely when there
is an IMBE so now they are now annotated with both so that
they are exported by at least/most one of the exporters.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/AbstractMessageHandler.java:
```diff
@@ -71,13 +86,19 @@ public final void handleMessage(Message<?> message) {
 				message = MessageHistory.write(message, this, this.getMessageBuilderFactory());
 			}
+			if (this.countsEnabled) {
+				this.handlerMetrics.beforeHandle(message);
```
`start` should be adjusted to the `System.currentTimeMillis()` here.
Otherwise in the `SimpleMessageHandlerMetrics` `this.duration.append(System.currentTimeMillis() - 0);` is very big value :smile:

artembilan replied:
OK. I see the code. Should be:
```
start = this.handlerMetrics.beforeHandle(message);
```

garyrussell replied:
doh

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/handler/management/SimpleMessageHandlerMetrics.java:
```diff
@@ -54,83 +47,44 @@
-
-			invocation.proceed();
+		this.handleCount.incrementAndGet();
+		this.activeCount.incrementAndGet();
```
Can you explain the difference between `handleCount` and `activeCount`.
I see that `activeCount` isn't resetted in the `reset()`.
But can't we adjust `activeCount` only on the `reset()` and add with the `handleCount` on the `getActiveCount()` call ?
I mean avoid this additional `incrementAndGet` on each `beforeHandle`

garyrussell replied:
`activeCount` is the number of currently active handles - as you can see it was previously decremented in the finally block; we can't do that now, but I need to add a decrement here (this method is always called after the handler handles the message).

Good catch; thanks.

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/history/MessageHistoryConfigurer.java:
```diff
@@ -45,7 +45,7 @@
  * @since 2.0
  */
-@ManagedResource
+@IntegrationManagedResource
```
And why have you removed `@ManagedResource` here ?
I may not be interested in handlers/channels metrics, but want to manage this component from JMX too

artembilan commented on spring-integration-core/src/main/java/org/springframework/integration/support/management/MetricsEnablement.java:
```diff
@@ -0,0 +1,37 @@
+ * @since 4.2
+ *
+ */
+public interface MetricsEnablement extends CountsEnablement{
```
Whitespace before `{`

artembilan commented on spring-integration-groovy/src/main/java/org/springframework/integration/groovy/config/GroovyControlBusFactoryBean.java:
```diff
@@ -16,22 +16,27 @@
 import org.springframework.util.CustomizableThreadCreator;
 
+import groovy.lang.Binding;
+import groovy.lang.MissingPropertyException;
```
Your name to the author list and reformat code, please.
The main concern - 120 line length.
And I'd prefer have blank line around method definitions even in the inline interface implementations.
I mean that `new ScriptVariableGenerator`

garyrussell replied:
I will add the author and blank lines, but I think I have explained before that I do not like reformatting just for the sake of it - it disturbs the git blame, making it hard to see why some change was made (it hides the real blame).

I prefer to fix formatting only when the code in violation requires a change for the feature/bug.

artembilan commented on spring-integration-jdbc/src/main/java/org/springframework/integration/jdbc/JdbcMessageStore.java:
```diff
@@ -77,10 +78,12 @@
  * @since 2.0
  */
 @ManagedResource
+@IntegrationManagedResource
```
These are redundant, since `AbstractMessageGroupStore` has both already

artembilan commented on spring-integration-jms/src/main/java/org/springframework/integration/jms/PollableJmsChannel.java:
```diff
@@ -65,6 +100,10 @@ public void setMessageSelector(String messageSelector) {
 			}
+			if (isCountsEnabled()) {
+				getMetrics().afterReceive();
+				counted = true;
```
Why is it here, rather than outside of `if`, like in the `PollableAmqpChannel` ,

garyrussell replied:
The amqp one should be inside the if too. Fixed.

artembilan commented on spring-integration-jmx/src/main/java/org/springframework/integration/jmx/config/MBeanExporterHelper.java:
```diff
@@ -84,16 +71,8 @@ public void afterPropertiesSet() throws Exception {
-					}
-					else if (className.equals(IntegrationMBeanExporter.class.getName())) {
-						this.hasIntegrationExporter = true;
+					if (className.startsWith(SI_ROOT_PACKAGE) && !(className.endsWith(IntegrationMBeanExporter.class.getName()))){
```
Looks like we can use here `IntegrationConfigUtils.BASE_PACKAGE` and remove both constants in this class.

artembilan commented on spring-integration-jmx/src/main/java/org/springframework/integration/monitor/IntegrationMBeanExporter.java:
```diff
@@ -111,13 +117,13 @@
  */
 @ManagedResource
 public class IntegrationMBeanExporter extends MBeanExporter implements BeanPostProcessor, BeanFactoryAware,
```
Why isn't this as `@IntegrationManagedResource` too ?

garyrussell replied:
Because the `IMBE` doesn't export itself.

http://docs.spring.io/spring-integration/reference/html/system-management-chapter.html#jmx-mbean-exporter

artembilan commented on spring-integration-jmx/src/main/java/org/springframework/integration/monitor/IntegrationMBeanExporter.java:
```diff
@@ -723,15 +719,23 @@ private void registerChannels() {
 
 	private void registerHandlers() {
-		for (SimpleMessageHandlerMetrics source : handlers) {
+		for (MessageHandlerMetrics source : handlers) {
```
s/source/handler

artembilan commented on spring-integration-jmx/src/main/java/org/springframework/integration/monitor/IntegrationMBeanExporter.java:
```diff
@@ -216,6 +221,16 @@ public void setComponentNamePatterns(String[] componentNamePatterns) {
+		this.enabledCountsPatterns = Arrays.copyOf(enabledCountsPatterns, enabledCountsPatterns.length);
+	}
+
+	public void setEnabledStatsPatterns(String[] enabledStatsPatterns) {
```
Need JavaDocs to these methods

artembilan commented:
In addition to those concerns there is need to add Docs and Migration Guide paragraph. However they can be done in the separate JIRA

garyrussell commented:
https://jira.spring.io/browse/INT-3651

garyrussell committed:
Polishing; PR Comments; Add Handler Metrics Test

Also found a problem using an inner MessageSource bean in an
inbound-channel-adapter - a BeanComponentDefinition was returned
instead of a BeanDefinition.

garyrussell committed:
Add @DirtiesContext to JMX Tests

The new MonitorTests failed with InstanceAlreadyExistsException for the
errorChannel.

Add @DirtiesContext to all tests using the Spring test runner to avoid caching any
contexts after the test class completes.

garyrussell commented:
Pushed - added dirties context annotations.

garyrussell committed:
INT-3637: JMX Support Negated Name Match Patterns

Note: with negated matches, order matters.

garyrussell commented:
Pushed

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/09c203dc6d37853763c5928c698d1aad4b339903

The :beers: time!
